File: scripts/compute.R

library(igraph, lib.loc="../../../../home/lim/R/x86_64-pc-linux-gnu-library/4.2/")
library(tidyverse, lib.loc="/home/lim/R/x86_64-pc-linux-gnu-library/4.2/")
library(jsonlite, lib.loc="/home/lim/R/x86_64-pc-linux-gnu-library/4.2/")

rm(list = ls())
# edges <- read.table("network.tsv", header = T, stringsAsFactors = F)[, 3:4]
edges <- read.table("https://string-db.org/api/tsv/network?identifiers=TP53", header = T, stringsAsFactors = F)[, 3:4]

nodes <- unique(c(edges[, 1], edges[, 2]))

net <- graph_from_data_frame(d = edges, vertices = nodes, directed = F)

degree <- data.frame("name" = names(degree(net, V(net))), "value" = degree(net, V(net)))
betweenness <- data.frame("name" = names(betweenness(net, V(net))), "value" = betweenness(net, V(net)))
closeness <- data.frame("name" = names(closeness(net, V(net))), "value" = closeness(net, V(net)))

para <- merge(degree, betweenness, by = "name") %>% 
  merge(closeness, by = "name")

names(para) <- c("Symbol", "Degree", "Betweenness", "Closeness")

parajson <- toJSON(para, pretty=TRUE)
print(parajson)

# cat(parajson, file = (con <- file('json.txt', "w", encoding = "UTF-8")))
# close(con)

# write.csv(para, "para.csv", row.names = F)

File: scripts/compute.py
import json
import igraph
import urllib.request
import pandas as pd
import sys

# url = "https://string-db.org/api/json/network?identifiers=TP53"

url = sys.argv[1]
data = urllib.request.urlopen(url)
text = json.load(data)

edgelist = []
for i in text:
    edgelist.append((i["preferredName_A"], i["preferredName_B"]))

edgelist = list(set(edgelist))

g = igraph.Graph.TupleList(edgelist, directed=False)
symbol = g.vs["name"]
degree = g.degree(symbol)
betweenness = g.betweenness(symbol)
closeness = g.closeness(symbol)

df = pd.DataFrame(list(zip(symbol, degree, betweenness, closeness)),
                  columns=["Symbol", "Degree", "Betweenness", "Closeness"])
df_html = df.to_html(index=False, table_id="topopara", border=0)
print(df_html)
